### PR TITLE
Add ability to enter explanation for work history breaks

### PR DESCRIPTION
**Pull request by chubberlisk**
### Context

For safeguarding reasons, a candidate should explain any break longer than a month in their work history. This option should only show if there is a gap in their work history.

### Changes proposed in this pull request

This PR adds to the ability to enter an explanation for work history breaks. It does so by updating `WorkHistoryReviewComponent` to show a summary card to enter this.

The responsibility of working out whether or not to show has been delegated to a new service called `CheckBreaksInWorkHistory`.

#### One job and no gap between end date and current date

image

#### One job and gap between end date and current date

image

#### Two jobs and gap between them

image

### Guidance to review

Best to go through each each commit.

Would particularly like feedback on the `CheckBreaksInWorkHistory` service and if there are any edge cases that are missing - see https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/482/commits/e14a7cbbe5725ea27888d14f9cac31fc01c9906f.

### Link to Trello card

[284 - Work history gaps](https://trello.com/c/MXo9XLXn/284-work-history-gaps)

### CODE REVIEW

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        expect(break_in_work_history).to eq(false)
+      end
+    end
+
+    context 'when there is only one job' do
+      it 'returns false if the job ends at current date' do
+        Timecop.freeze(october2019) do
+          application_form = create(:application_form) do |form|
+            form.application_work_experiences.create(
+              start_date: november2019,
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: This should be `september2019`.

File: app/controllers/candidate_interface/work_history/breaks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        redirect_to candidate_interface_work_history_show_path
+      else
+        render :edit
+      end
+    end
+
+  private
+
+    def work_breaks_form_params
+      return nil unless params.has_key?(:candidate_interface_work_breaks_form)
```
Comment (chubberlisk) [inline app/controllers/candidate_interface/work_history/breaks_controller.rb]: Is this needed?

File: app/services/check_breaks_in_work_history.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+class CheckBreaksInWorkHistory
+  class << self
+    def call(application_form)
+      jobs = application_form.application_work_experiences.sort_by(&:start_date)
+
+      return false if jobs.empty?
+      return breaks_between_only_job_and_current_date?(jobs.first) if jobs.one?
+
+      jobs.each_cons(2).any? do |first_job, next_job|
+        month_or_more_between?(first_job, next_job)
```
Comment (chubberlisk) [inline app/services/check_breaks_in_work_history.rb]: `month_or_more_break_between?`

File: app/services/check_breaks_in_work_history.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+class CheckBreaksInWorkHistory
+  class << self
+    def call(application_form)
+      jobs = application_form.application_work_experiences.sort_by(&:start_date)
+
+      return false if jobs.empty?
+      return breaks_between_only_job_and_current_date?(jobs.first) if jobs.one?
```
Comment (chubberlisk) [inline app/services/check_breaks_in_work_history.rb]: `break_between_only_job_and_current_date?`

File: app/models/candidate_interface/work_breaks_form.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+module CandidateInterface
+  class WorkBreaksForm
```
Comment (chubberlisk) [inline app/models/candidate_interface/work_breaks_form.rb]: `WorkHistoryBreaksForm`

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    let(:march2019) { Time.zone.local(2019, 3, 1) }
+    let(:june2019) { Time.zone.local(2019, 6, 1) }
+    let(:august2019) { Time.zone.local(2019, 8, 1) }
+    let(:september2019) { Time.zone.local(2019, 9, 1) }
+    let(:october2019) { Time.zone.local(2019, 10, 1) }
+    let(:november2019) { Time.zone.local(2019, 11, 1) }
+
+    context 'when there are no jobs' do
+      it 'returns false' do
+        application_form = create(:application_form)
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: Can be `build` or `builfld_stubbed`.

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            )
+          end
+
+          break_in_work_history = CheckBreaksInWorkHistory.call(application_form)
+
+          expect(break_in_work_history).to eq(false)
+        end
+      end
+
+      it 'returns false if the job is current role i.e. end date is blank' do
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: Is nil

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        Timecop.freeze(october2019) do
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: Not needed.

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            )
+          end
+
+          break_in_work_history = CheckBreaksInWorkHistory.call(application_form)
+
+          expect(break_in_work_history).to eq(false)
+        end
+      end
+
+      it 'returns true if the job ends a month before current date' do
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: Need a test for "or more"

File: spec/services/check_breaks_in_work_history_spec.rb [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            end_date: october2019,
+          )
+        end
+
+        break_in_work_history = CheckBreaksInWorkHistory.call(application_form)
+
+        expect(break_in_work_history).to eq(false)
+      end
+
+      it 'returns false if there are breaks in work history' do
```
Comment (chubberlisk) [inline spec/services/check_breaks_in_work_history_spec.rb]: Returns true

File: spec/system/candidate_interface/candidate_entering_work_history_spec.rb [hunk 11]
```diff
@@ -28,6 +28,17 @@
     and_i_fill_in_the_job_form
     then_i_should_see_my_completed_job
 
+    when_i_click_on_add_another_job
+    and_i_fill_in_the_job_form_with_another_job
```
Comment (chubberlisk) [inline spec/system/candidate_interface/candidate_entering_work_history_spec.rb]: With a break

File: app/services/check_breaks_in_work_history.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+class CheckBreaksInWorkHistory
```
Comment (benilovj) [inline app/services/check_breaks_in_work_history.rb]: There is lots of logic in this class that IMHO better suites to living in a`WorkHistory` (class or something similarly named) which would operate over the collection of work experiences. Then the service could be very thin (or disappear completely).

Comment (benilovj) [reply app/services/check_breaks_in_work_history.rb]: Given where we are in the delivery, feel free to ignore.

**Pull request merged**